### PR TITLE
Keep the DeepSeek router GEMM in fp32 under deterministic inference

### LINKED ISSUE #34758: [Feature] Router GEMM should keep fp32 output under deterministic inference (DeepSeek V3/V4)
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

The DeepSeek router GEMM should output fp32 under deterministic inference too. Today it falls back to bf16, and that is not needed for batch invariance.

`python/sglang/srt/models/deepseek_v2.py:512`:

```python
if get_exec().deterministic.enable_deterministic_inference:
    return F.linear(hidden_states, self.weight, None)   # -> bf16 logits
```

Every other branch in the same function returns fp32: `dsv3_router_gemm(out_dtype=torch.float32)`, `linear_bf16_fp32`, and the prefill-CP path.

**fp32 output does not break batch invariance.** Batch invariance is about reduction order, not output dtype. Deterministic `F.linear` runs `matmul_kernel_persistent`, which accumulates in fp32 and already has an fp32 store branch (`batch_invariant_ops.py:167`) that is never reached. It is the same kernel with the same reduction; only the final store differs. The branch is unreachable because `_matmul_persistent_triton` allocates output as the input dtype (`batch_invariant_ops.py:185`). Same for the JIT `dsv3_router_gemm` kernel, where `kOutFloat` only changes `cast<OutT>(final_sum)`.

**Impact.** Router logits feed `noaux_tc` top-k over 256/384 experts with an fp32 correction bias. bf16 has 8 mantissa bits, so adjacent expert scores often sit within one ULP and rounding can change which experts are selected — a different answer, not a small numeric difference.

**Related bug.** `_mm_dtype_compat` (`batch_invariant_ops.py:967`), registered for `aten::mm.dtype`, does `matmul_persistent(...).to(out_dtype)`, i.e. fp32 -> bf16 -> fp32. Callers asking for fp32 get bf16 values.

### Proposed change

1. Add `out_dtype` (or a preallocated `output`) to `matmul_persistent` / `_matmul_persistent_triton` so the existing fp32 store branch is reachable. Tiling and accumulation order are unchanged.
2. Use it in `_mm_dtype_compat` instead of `.to(out_dtype)`.
3. In `DeepseekV2MoEGate.forward`, request fp32 out from the batch-invariant matmul instead of returning bare `F.linear`.
4. Keep one kernel for all M under deterministic mode — the normal path switches kernels at `max_router_gemm_tokens` (4 on SM100/103, else 16), and a token-count-dependent switch is batch-variant on its own.
5. `_matmul_persistent_deepgemm` allocates `out` as `a.dtype` for `deep_gemm.bf16_gemm_nn`; either find an fp32-out entry point or route fp32 requests to Triton.

Cost should be negligible — the GEMM is `[M, 7168] x [7168, 256]` and this only changes the store dtype.

### Related resources

- `python/sglang/srt/models/deepseek_v2.py:512`
- `python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py:167,185,275,967`
- `python/sglang/kernels/jit/csrc/gemm/dsv3_router_gemm.cuh` — `kOutFloat`
- #12095 — added the `F.linear` fallback

### PR DESCRIPTION
**Pull request by pol-resina**
## Motivation

Closes #34758.

Under deterministic inference the DeepSeek router GEMM returns bf16 logits, while every other branch of `MoEGate.forward` returns fp32 (`dsv3_router_gemm(out_dtype=torch.float32)`, `linear_bf16_fp32`, the prefill-CP path). `noaux_tc` ranks 256/384 experts against an fp32 correction bias, and adjacent expert scores routinely sit within one bf16 ULP, so the narrower store changes which experts run — a different answer, not a small numeric difference.

fp32 output does not cost batch invariance. Invariance is a property of the reduction order, and the store happens after the reduction is complete.

## Modifications

The root cause is one missing degree of freedom: `matmul_persistent` could not express *accumulate fp32, store fp32* — the output dtype was implicitly `a.dtype`. Every symptom in the issue follows from that.

**`batch_invariant_ops.py`**

- `out_dtype` threaded through `matmul_persistent`, `_matmul_persistent_triton` and `_matmul_persistent_deepgemm`.
- The Triton tile config stays keyed on the **input** dtype. This is load-bearing: the configs are not the same shape (`bfloat16 -> BLOCK_SIZE_K=64`, `float32 -> BLOCK_SIZE_K=32`), so keying on `out_dtype`
- would halve the k-tile and silently run a different reduction than the non-deterministic path. Only the `torch.empty` moves.
- DeepGEMM keeps its kernel instead of falling back to Triton. `bf16_gemm_nn` only stores bf16, but 
- bf16_gemm_nt` already takes an fp32 `out` (it is what `gemm_nt_bf16bf16f32` and `_BF16F32WarmupExecutor` use) and wants the `[N, K]` rhs that `matmul_persistent` has already checked is contiguous. Routing fp32 to Triton would have demoted the kernel on exactly the H200/B200 deterministic runs this is meant to help.
- `_mm_dtype_compat` asks the kernel for `out_dtype` rather than casting afterwards, fixing the `fp32 -> bf16 -> fp32` round trip that handed `aten::mm.dtype` callers bf16 values.
- Its two `.contiguous()` calls are dropped. For the router, `mat2` is `weight.t()`, so they were materializing a `[7168, 256]` copy every forward, every MoE layer, and disqualifying the DeepGEMM path on the way. Its sibling `mm_batch_invariant` never did this, and the kernel takes strided operands.

**`deepseek_v2.py`** — the deterministic branch becomes `torch.mm(hidden_states, self.weight.t(), out_dtype=torch.float32)`. Going through the aten override rather than importing `matmul_persistent` keeps  device dispatch intact. The issue's fourth point needs no work: this branch already returns above the `max_router_gemm_tokens` switch, so the token-count-dependent kernel choice was never reachable under deterministic mode.

  **`npu_batch_invariant_ops.py`** — `aten::mm.dtype` registered on NPU. It casts rather than widening (the NPU kernel stores its input dtype), so precision there is unchanged; without it `out_dtype=` would fall
  through to the stock NPU matmul and leave batch-invariant mode altogether.

## Accuracy Tests

Measured on one GPU with the Triton backend, `[64, 7168] x [7168, 256]` shaped like the router GEMM.

  
The reduction is untouched — the bf16 result is bit-identical to the fp32 result rounded:

  ```
  bf16 store == round(fp32 store)          True 
  distinct values, bf16 / fp32             1958 / 16382  of 16384
  max abs err vs fp64   bf16 9.800e-01  ->  fp32 3.895e-03   (~250x)
  batch invariant for M in {1,2,4,5,16,17,64}   exact
  ```
Expert-selection impact, holding the accumulator fixed and changing only the store, over `[4096, 7168] x [7168, 256]` with top-8 and an fp32 correction bias:

  | router logit std | tokens with a different expert set |
  |---|---|
  | 0.05 | 18 / 4096 |
  | 0.10 | 54 / 4096 |
  | 0.25 | 77 / 4096 |
  | 0.50 | 90 / 4096 |
  | 1.00 | 109 / 4096 |
  | 2.00 | 84 / 4096 |

0.4%–2.7% of tokens routed to a different expert set purely from the bf16 store. These are synthetic weights; a real DeepSeek-V3 comparison would be stronger and I do not have an 8xH200 to run it on.

New tests in `test/registered/unit/batch_invariant_ops/test_matmul_out_dtype.py`, registered `base-b` / `1-gpu-large` rather than `1-gpu-small`: DeepGEMM is disabled on SM120, so only the H100 runner exercises
  the fp32 `bf16_gemm_nt` path. The file sets `_ENABLE_MM_COMPARISON_TEST`, and the operands satisfy the DeepGEMM gate, so CI diffs NT against Triton on every push.

## Speed Tests and Profiling

Not benchmarked. The GEMM is `[M, 7168] x [7168, 256]` and the kernel choice is unchanged on every backend, so the only cost is the wider store; the `.contiguous()` removal should be a small win on the deterministic path.

## Notes for reviewers

Two things worth a closer look:

1. The `bf16_gemm_nt` fp32-out branch is reasoned from `gemm_nt_bf16bf16f32` and `_BF16F32WarmupExecutor`, not executed locally — no DeepGEMM on the hardware I have. The new test covers it on the first H100 CI
run.
2. The `.contiguous()` removal in `_mm_dtype_compat` is the one change beyond the literal ask.

## Checklist

  - [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
  - [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
  - [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
  - [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the 
  speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
  - [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
  
## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
     - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #34717350396](https://github.com/sgl-project/sglang/actions/runs/34717350396)
Latest PR Test (Extra): :x: [Run #34717350322](https://github.com/sgl-project/sglang/actions/runs/34717350322)
Latest PR Test (AMD ROCm 10): :x: [Run #34717350394](https://github.com/sgl-project/sglang/actions/runs/34717350394)

### CODE REVIEW

**Comment by pol-resina** [comment]
Rebased onto current main (was ~750 commits behind, which was failing check-changes via the MIN_BASE_SHA gate). Could someone run /tag-and-rerun-ci so CI runs on the current commit? cc @b8zhong. This implements #34758.

File: python/sglang/srt/hardware_backend/npu/batch_invariant_ops/npu_batch_invariant_ops.py [hunk 1]
```diff
@@ -9,6 +9,12 @@ def npu_mm_batch_invariant(a, b):
     return torch.ops.batch_invariant_ops.npu_mm_batch_invariant(a, b)
 
 
+def npu_mm_dtype_batch_invariant(a, b, out_dtype):
+    # A cast, not a wider store, but it keeps out_dtype= off the stock NPU matmul.
+    out = torch.ops.batch_invariant_ops.npu_mm_batch_invariant(a, b)
+    return out if out.dtype == out_dtype else out.to(out_dtype)
```
Comment (xikronz) [inline python/sglang/srt/hardware_backend/npu/batch_invariant_ops/npu_batch_invariant_ops.py:15]: @b8zhong should we include a NPU kernel for writing GEMM in fp32 here so that the router-logit fix is properly supported across all backends or just leave this PRs scope for CUDA devices only? the current NPU handler preserves batch invariance but only widens an already-rounded bf16 result (resurfaces the same issue as before)

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py [hunk 2]
```diff
@@ -304,15 +325,16 @@ def matmul_persistent(
...
-        return _matmul_persistent_deepgemm(a=a, b=b, bias=bias)
+        return _matmul_persistent_deepgemm(a=a, b=b, bias=bias, out_dtype=out_dtype)
 
     if _ENABLE_MM_FALLBACK_VARIANT:
         out = torch.einsum("ik,kj->ij", a, b)
         if bias is not None:
             out += bias
-        return out
+        # Debug-only path: the variant GEMM still stores a.dtype, so this is a cast.
+        return out if out_dtype is None else out.to(out_dtype)
```
Comment (xikronz) [inline python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: not sure if labeling this path as “debug-only” makes the post-cast correct. `_ENABLE_MM_FALLBACL_VARIANT` is an opt-in fallback for shapes DeepGEMM cant handle... here einsum still stores GEMM outputs into a bf16 buffer before calling .to(torch.float32) which reproduces the precision loss this PR is trying to fix. maybe restrict the condition before entering this block?

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py [hunk 3]
```diff
@@ -248,15 +253,22 @@ def grid(META):
...
     M, K = a.shape
     K, N = b.shape
-    dtype = a.dtype
-    out = torch.empty((M, N), device=a.device, dtype=dtype)
+    out_dtype = a.dtype if out_dtype is None else out_dtype
+    out = torch.empty((M, N), device=a.device, dtype=out_dtype)
 
     try:
-        deep_gemm.bf16_gemm_nn(a, b, out)
+        if out_dtype == torch.float32:
```
Comment (xikronz) [inline python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: nit: in the pinned DeepGEMM v0.1.5 implementation, bf16_gemm_nn(a, b, out) already delegates to bf16_gemm_nt(a, b.transpose(0, 1), out) and the NT implementation accepts either bf16 or fp32 as output soo the conditional here is a bit redundant

File: test/registered/unit/batch_invariant_ops/test_matmul_out_dtype.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
+"""Batch-invariant matmul with an explicit output dtype.
```
Comment (xikronz) [inline test/registered/unit/batch_invariant_ops/test_matmul_out_dtype.py:1]: this PR's changes also affects NPU/XPU backends but the tests cover only CUDA/AMD. could we add NPU/XPU coverage in here?

File: python/sglang/srt/hardware_backend/npu/batch_invariant_ops/npu_batch_invariant_ops.py (hunk 1 above)
Comment (pol-resina) [reply python/sglang/srt/hardware_backend/npu/batch_invariant_ops/npu_batch_invariant_ops.py:15]: Kernel lives in external package. Not registering is worse. @b8zhong call

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py (hunk 2 above)
Comment (pol-resina) [reply python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: Agreed, took your fix

File: python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py (hunk 3 above)
Comment (pol-resina) [reply python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py]: Agreed, took your fix

File: test/registered/unit/batch_invariant_ops/test_matmul_out_dtype.py (hunk 4 above)
Comment (pol-resina) [reply test/registered/unit/batch_invariant_ops/test_matmul_out_dtype.py:1]: XPU already covered via the generic Triton path. NPU cant run this file
